File: ios/DMRandevuGaleri/Media/Censor/BeepPlayer.swift
import AVFoundation

/// The censor tone, played live over the video while a marked stretch goes past.
///
/// So the operator can hear what they marked instead of reading a red bar and hoping. It is an
/// approximation of the export, not the export: the real one separates the voice out and leaves
/// the music underneath, which takes seconds a frame and cannot happen during playback. Here the
/// video is simply ducked and the tone laid over it, which is enough to tell whether the mark
/// covers the word — the question the operator is actually asking.
final class BeepPlayer {

    private var player: AVAudioPlayer?

    /// Starts the tone, or does nothing if it is already sounding.
    func start() {
        if player?.isPlaying == true { return }
        if player == nil { player = try? AVAudioPlayer(data: Self.tone()) }
        guard let player else { return }
        // Looped rather than one long buffer, so a mark of any length is covered.
        player.numberOfLoops = -1
        player.volume = Float(Self.level)
        player.prepareToPlay()
        player.play()
    }

    func stop() {
        player?.stop()
        player?.currentTime = 0
    }

    /// One second of 1 kHz as a WAV, eased at both ends so looping it does not tick.
    private static func tone() -> Data {
        let count = sampleRate
        let fade = sampleRate * fadeMs / 1000
        var samples = [Int16](repeating: 0, count: count)
        for i in 0..<count {
            let envelope: Double
            if i < fade {
                envelope = 0.5 - 0.5 * cos(Double.pi * Double(i) / Double(fade))
            } else if i > count - fade {
                envelope = 0.5 - 0.5 * cos(Double.pi * Double(count - i) / Double(fade))
            } else {
                envelope = 1
            }
            let value = envelope * sin(2 * Double.pi * PcmOps.beepHz * Double(i) / Double(sampleRate))
            samples[i] = Int16(value * 32000)
        }
        return wav(samples)
    }

    /// The smallest WAV that AVAudioPlayer will accept: it takes data, not raw PCM.
    private static func wav(_ samples: [Int16]) -> Data {
        let bytes = samples.count * 2
        var data = Data()
        func append(_ string: String) { data.append(contentsOf: Array(string.utf8)) }
        func append32(_ value: UInt32) { withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) } }
        func append16(_ value: UInt16) { withUnsafeBytes(of: value.littleEndian) { data.append(contentsOf: $0) } }

        append("RIFF")
        append32(UInt32(36 + bytes))
        append("WAVE")
        append("fmt ")
        append32(16)
        append16(1)                              // PCM
        append16(1)                              // mono
        append32(UInt32(sampleRate))
        append32(UInt32(sampleRate * 2))         // bytes a second
        append16(2)                              // bytes a frame
        append16(16)                             // bits a sample
        append("data")
        append32(UInt32(bytes))
        samples.withUnsafeBufferPointer { data.append(UnsafeRawBufferPointer($0).bindMemory(to: UInt8.self)) }
        return data
    }

    private static let sampleRate = 44_100
    private static let fadeMs = 5

    /// Quieter than the exported beep, which is not competing with a ducked video.
    private static let level = 0.18
}

File: app/src/main/java/com/dmrandevu/gallery/media/censor/BeepPlayer.kt
package com.dmrandevu.gallery.media.censor

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/**
 * The censor tone, played live over the video while a marked stretch goes past.
 *
 * So the operator can hear what they marked instead of reading a red bar and hoping. It is an
 * approximation of the export, not the export: the real one separates the voice out and leaves
 * the music underneath, which takes seconds a frame and cannot happen during playback. Here the
 * video is simply ducked and the tone laid over it, which is enough to tell whether the mark
 * covers the word — the question the operator is actually asking.
 */
class BeepPlayer {

    private var track: AudioTrack? = null

    /** Starts the tone, or does nothing if it is already sounding. */
    fun start() {
        if (track != null) return
        val samples = tone()
        val track = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build()
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setSampleRate(SAMPLE_RATE)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build()
            )
            .setBufferSizeInBytes(samples.size * 2)
            .setTransferMode(AudioTrack.MODE_STATIC)
            .build()
        track.write(samples, 0, samples.size)
        // Looped over a whole number of cycles, so the join is silent.
        track.setLoopPoints(0, samples.size, -1)
        track.play()
        this.track = track
    }

    fun stop() {
        track?.let {
            runCatching { it.stop() }
            it.release()
        }
        track = null
    }

    /** One second of 1 kHz, eased at both ends so looping it does not tick. */
    private fun tone(): ShortArray {
        val count = SAMPLE_RATE
        val fade = SAMPLE_RATE * FADE_MS / 1000
        return ShortArray(count) { i ->
            val envelope = when {
                i < fade -> 0.5 - 0.5 * cos(PI * i / fade)
                i > count - fade -> 0.5 - 0.5 * cos(PI * (count - i) / fade)
                else -> 1.0
            }
            (LEVEL * envelope * sin(2 * PI * PcmOps.BEEP_HZ * i / SAMPLE_RATE) * 32767).toInt()
                .toShort()
        }
    }

    private companion object {
        const val SAMPLE_RATE = 44_100
        const val FADE_MS = 5

        /** Quieter than the exported beep, which is not competing with a ducked video. */
        const val LEVEL = 0.18
    }
}
